pkg/controller/cloud/node_controller.go

```diff
@@ -337,6 +337,21 @@ func (cnc *CloudNodeController) AddCloudNode(obj interface{}) {
 	}
 
 	err := clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
+		// TODO(wlan0): Move this logic to the route controller using the node taint instead of condition
```

Comment:
What's the justification of moving this here earlier? Is it to set the node condition as early as possible?

cc @cheftako

Reply:
Yes, we must ensure that `NodeNetworkUnavailable` condition is added before we remove the cloud taint.

Btw, by moving these code ahead of the `curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})`, the curNode's would have `NodeNetworkUnavailable`  condition after we get it from apiserver by `cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})`.
If we don't move these code, when we update the node condition with `nodeutil.SetNodeCondition` methods, the curNode's status still doesn't have `NodeNetworkUnavailable`  condition, which is a little inconsistent, and we need to `Get` the node from apiserver again.

Reply:
Not sure I follow, the current implementation is 
```
curNode.Status.Conditions = append(node.Status.Conditions, v1.NodeCondition{
				Type:               v1.NodeNetworkUnavailable,
				Status:             v1.ConditionTrue,
				Reason:             "NoRouteCreated",
				Message:            "Node created without a route",
				LastTransitionTime: metav1.Now(),
			})
```
Which updates the condition on the node object in memory but does not apply any changes to the apiserver yet. The current implementation bundles all the necessary updates (removing taints, conditions, zones, instance types, etc) into 1 update call [here](https://github.com/kubernetes/kubernetes/pull/65052/files#diff-a1ff61328540d76c791d919f71da4ad5R409) so there is no need to call `nodeutil.SetNodeCondition` as far as I know.

Reply:
Is it OK to remove this condition before we know the network is actually ok? I think we need someone from GCP to review this as well.

Reply:
@andrewsykim 

The one update call doesn't update the node condition actually, the reason was describe here. #65051

Reply:
This is good to know, thanks :)

Reply:
@dbdd4us, @andrewsykim 
I think it is correct to move the no route condition earlier. I'm slightly worried about the earlier error exit we are adding. We seem to have several such early error exits, so nothing in this block are guaranteed to happen. I don't think this is particular branch is actually exercised today. I should be testing this well before it is enabled on the main line. Given that I am ok with the change.